symengine/dict.cpp

```diff
@@ -104,9 +105,64 @@ std::ostream& operator<<(std::ostream& out, const SymEngine::set_basic& d)
     return print_vec_rcp(out, d);
 }
 
+#if defined(HAVE_SYMENGINE_PIRANHA)
+std::ostream& operator<<(std::ostream& out, const SymEngine::hash_set& d)
+{
+    return print_map(out, d);
+}
+#endif
 
 namespace SymEngine {
 
+#if defined(HAVE_SYMENGINE_PIRANHA)
+bool hash_set_eq(const hash_set &a,
+        const hash_set &b)
+{
+    // Can't be equal if # of elements in set differ:
+    if (a.size() != b.size()) return false;
+    // Loop over elements in "a":
+    for (auto &p: a) {
+        auto f = b.find(p);
+        if (f == b.end()) return false; // no such element in "b"
+    }
+    return true;
+}
+
+#else
+bool unordered_set_eq(const unordered_set &a,
+        const unordered_set &b)
+{
+    // Can't be equal if # of elements in set differ:
+    if (a.size() != b.size()) return false;
+    // Loop over elements in "a":
+    for (auto &p: a) {
+        auto f = b.find(p);
+        if (f == b.end()) return false; // no such element in "b"
+    }
+    return true;
+}
```

Comment:
Implement `set_eq` as a template, similar to `umap_eq`: https://github.com/sympy/symengine/blob/b873693e362981ec664f1411e030a7b170c79d3a/symengine/dict.h#L50 and then just call `set_eq` from `hash_set_eq` and from `unordered_set_eq`. In fact, as it goes, you probably do not even need to define the two tiny wrappers and just call `set_eq` directly instead of `hash_set_eq`. You do not need to specify the templates when you call it, the compiler should be able to figure it out.